### DISCUSSION TITLE
Bulk Import of Preview pictures

### DISCUSSION THREAD

**Discussion by RaptorDE** [Part-DB/Part-DB-server#439] [Q&A] [answered]
I have an existing spare parts database that I would like to migrate to part-db. 
I have already been able to import my over 1000 data sets very well using the CSV import function. Unfortunately, I haven't found a way to import the associated preview images. I have a list in which I can reference the file names of the images with the IPN. Is there a way to import these images? Possibly by manually editing the database or with the API? 
I tried the POST /api/attachments function, unfortunately I couldn't add an image.

Request body:
```json
{
  "name": "Vorschaubild1",
  "element": {
    "name": "Vorschaubild1"
  },
  "url": "/media/import/rc_part_image_00a10f6a-0e0e-41c3-94ad-3eadbbafdc3d.png"
}
```
Response body:
```json
{
  "@context": "/api/contexts/Error",
  "@id": "/api/errors/500",
  "@type": "hydra:Error",
  "title": "An error occurred",
  "detail": "Cannot instantiate abstract class App\\Entity\\Attachments\\Attachment",
  "status": 500,
  "type": "/errors/500",
  "hydra:title": "An error occurred",
  "hydra:description": "Cannot instantiate abstract class App\\Entity\\Attachments\\Attachment"
}
```

**Comment by jbtronics** [comment]
There currently is no way to upload files via the API and for security reason it is not possible to modify the underlying file pathes directly.

I guess the easiest approach would be to directly manipulating the database and adding new rows in the attachment table.

**Reply by RaptorDE** [reply]
Thank you very much for the quick answer, I'll try that. 
Apart from a data backup and working at my own risk, is there anything else that needs to be taken into account when editing the SQLite database? After editing, should I clear the cache with `partdb:cache:clear` and check the database with `doctrine:migrations:up-to-date` and `doctrine:migrations:migrate`?

**Reply by jbtronics** [reply]
As long as you dont change the table structure itself and just create new rows, clearing the cache should be enough.

You should just try to avoid having two attachments with the same name on the same element, as this is normally forbidden.

**Reply by RaptorDE** [reply, marked as answer]
I have successfully edited the database to add the preview images. Using the HeidiSQL software, I was able to easily export the attachments table as a CSV and add the new attachments in Excel. I then imported the table again. I made two errors on my first try. By editing in Excel, the date format for the entries was no longer correct and I had to restore the Linux permission of the database file for the user www-data. Thanks for your support.

**Comment by RaptorDE** [comment]
It is now possible to upload files via the API. 🥳 
Here is my quickly cobbled together Python script for import.

In order to connect the images to the parts, the parts must be favorite and have an IPN whose number is the same as the file name in the images subfolder. E.g. IPN = EBN10 File name = Image_10.jpg The ID of the attachment_type is 1002. But everything can also be adjusted in the script.


> [!CAUTION]
> Make a backup before use!

```

import requests
import base64
import os

# API-Endpunkt und Header
base_url = 'http://[Part-DB-IP]:8080/api' # Ersetze durch deine Server URL

headers = {
    'Authorization': 'Bearer [API-KEY]', # Ersetze durch dein Token
    'accept': 'application/ld+json'
}

# Funktion zum Abrufen der Part-ID anhand des IPN
def get_part_id(ipn):
    search_url = f"{base_url}/parts?ipn={ipn}&favorite=true"
    response = requests.get(search_url, headers=headers)
    
    if response.status_code == 200:
        parts = response.json()["hydra:member"]
        if parts:
            part_id = parts[0]['@id']  # Erster Treffer
            print(f"Gefundene Part-ID für {ipn}: {part_id}")
            return part_id
        else:
            print(f"Kein Teil gefunden für {ipn}.")
            return None
    else:
        print(f"Fehler bei der Anfrage für {ipn}:", response.status_code, response.text)
        return None

# Datei in Base64 kodieren
def encode_file_to_base64(filepath):
    with open(filepath, 'rb') as file:
        encoded = base64.b64encode(file.read()).decode('utf-8')
        return encoded

# Bilder im Verzeichnis verarbeiten
image_folder = 'images/'
for filename in os.listdir(image_folder):
    if filename.endswith(".jpg"):
        # IPN aus dem Dateinamen extrahieren (Name_nummer.jpg -> EBNnummer)
        name_parts = filename.rsplit("_", 1)  # Nur das letzte _ für die Nummer verwenden
        if len(name_parts) == 2:
            ipn = f"EBN{name_parts[1].split('.')[0]}"
            part_id = get_part_id(ipn)

            if part_id:
                # Bild hochladen
                filepath = os.path.join(image_folder, filename)
                data = {
                    "name": filename,
                    "attachment_type": "/api/attachment_types/1002",
                    "element": part_id,
                    "upload": {
                        "data": f"data:image/jpg;base64,{encode_file_to_base64(filepath)}",
                        "filename": filename,
                        "private": False
                    }
                }
                upload_url = f"{base_url}/attachments"
                response = requests.post(upload_url, json=data, headers=headers)

                if response.status_code == 201:
                    print(f"Bild {filename} erfolgreich hochgeladen:", response.json())
                else:
                    print(f"Fehler beim Hochladen von {filename}:", response.status_code, response.text)
            else:
                print(f"Keine gültige Part-ID für {filename} gefunden.")


```
